Add 'apidoc_flag' to perlapi generation

khwilliamson opened the pull request:
This is designed to make noting what flag bits a function or macro takes very easy, while helping downstream code that processes things do a better job.

* This set of changes does not require a perldelta entry.

khwilliamson committed:
perlapi: Use new 'apidoc_flag' lines

This replaces the clumsy method used previously by the new construct
introduced in the previous commit.

There are many other potential places to use this new feature; this gets
just the important ones (the ones having separate apidoc entries for the
flags) and a few others that I tried.  The others aren't important
currently, as the only current effect of adding them would be to add X<>
entries to perlapi; that could change.  Devel::PPPort can be enhanced to
take advantage of these as well.

tonycoz commented on scope.c:
```diff
@@ -291,8 +291,8 @@ magic will be processed on the new scalar.  If unset, 'set' magic will be
-=for apidoc Amnh ||SAVEf_KEEPOLDELEM
-=for apidoc Amnh ||SAVEf_SETMAGIC
+=for apidoc Cmnh ||SAVEf_KEEPOLDELEM
+=for apidoc Cmnh ||SAVEf_SETMAGIC
```
These are also used by and documented for save_helem_flags(), which is API.

khwilliamson committed:
embed.pl: Mv line to avoid a pattern match

We've already done the match by the time of the new place

khwilliamson committed:
embed.pl: White-space/comments only

khwilliamson committed:
embed.pl: Skip embedvar.h in calculating visibility

This header file is constructed by this program.  It has nothing to
offer to the portions of the program that calculate symbol visibilities.
(Nothing in it indicates visibility)  So, just skip it.

khwilliamson committed:
embed.pl: Default Perl reserveds are always visible

Symbols like PL_foo should be considered to be visible to any module
wanting to use it, unless there is a declaration to the contrary.

khwilliamson committed:
Fix up some intrpvar variable visibilities

Currently all PL_ variables are considered to be visible everywhere, but
a future commit will start to pay attention to their declared
visibilities (defaulting to always visible if no declaration).

This commit makes sure that will leave the visibilities as they
currently are by adding the ones declared to being hidden to the
unresolved override list.  I also looked at a snapshot of metacpan for
any uses of these symbols, and if found, instead changed their
declarations to be always visible.

khwilliamson committed:
embed.pl: Build up re's from smaller ones

I'm not sure this is worth bothering with, but this creates longer
string by adding to shorter one, and then constructs regex patterns from
the various steps.

khwilliamson committed:
embed.pl: Keep track of input line number for debug

I found myself wanting to know exactly where something is in the files
read by this program.

khwilliamson committed:
embed.pl: Get stored visibility if not explicit

We can have groups like
    apidoc      Am|foo
    apidoc_item    bar
    ...

This says the visibility of 'foo' is A (for all, or everywhere), and bar
inherits that.  But we can also have groups like
    apidoc      foo
    apidoc_item bar

where "foo's" visibility is defined elsewhere.  For example, "foo" could
actually be a function, and its visibility defined in embed.fnc.

In the latter case, we've already parsed embed.fnc, and so likely know
the visibility of 'foo' and have stored it.

This commit looks at the place it would be stored and uses that if there
is no visibility specified here for 'foo'.  This is somewhat cheating,
to know how the data structure is laid out, but it is the most
convenient way to do this, and I think, acceptable.

This single line change to the code pointed out several instances where
the ommission of this value led to not knowing what to do, so several
symbols were added to the override list that actually weren't a problem,
and so this commit removes them.

khwilliamson committed:
Add 'apidoc-flag' to perlapi generation

This is designed to make noting what flag bits a function or macro takes
very easy, while helping downstream code that processes things do a
better job.

khwilliamson committed:
embed.pl: Clarify code, add comment

Instead of omitting a parameter, explicitly mark it 'undef' for code
readability, and add a comment as to what this is all about.

tonycoz commented on sv.c:
```diff
@@ -5130,7 +5130,11 @@ S_newSVsv_flags_NN_PVxx(pTHX_ SV* dsv, SV* ssv, const I32 flags)
+=for apidoc_flag SVs_TEMP
+=for apidoc_flag SV_NOSTEAL
+=for apidoc_flag SV_GMAGIC
+=for apidoc_flag SV_SMAGIC
```
I assume you copied SMAGIC since it's available for sv_setsv_flags(), but it is meaningless for a new SV.

Though sv_setsv_flags() doesn't directly document it takes SV_SMAGIC (or SV_GMAGIC).

But this isn't documentation, so I'm not sure it has any meaning.

khwilliamson replied:
Thanks.  You presumed correctly; I had skimmed what this calls, and saw those flags mentioned, without noticing that they are not used when called in this particular case.  And all that got generated here were inappropriate indexing calls X<>.  So I've removed these. 

 sv_setsv_flags() does have apidoc_flag lines for these flags; it doesn't use their names explicitly in its text, but the reader can readily infer that they are to be used, so I didn't add explicit mention of them.

The point of this p.r. is to make it easier to document that a call takes flags, superseding the clumsy previous method.  It could lead to automatic text at the bottom of entries saying "foo takes the X, Y, and Z flags"; I don't know.  But it also is useful for Devel::PPPort in making automatic test cases.
